codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,164 @@
+    /**
+     * Write array header only without body. Use this if you want to write arrays as streaming.
+     */
+    private static void writeArrayHeaderObject(ByteBuf buf, ArrayHeaderRedisMessage msg) {
+        buf.writeByte(RedisMessageType.ARRAY.value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+        } else {
+            buf.writeBytes(numberToBytes(msg.length()));
+        }
+        buf.writeBytes(CRLF);
+    }
+
+    /**
+     * Write full constructed array message.
+     */
+    private static void writeArrayMessage(ByteBufAllocator allocator, ByteBuf buf, ArrayRedisMessage msg) {
+        buf.writeByte(msg.type().value());
+        if (msg.isNull()) {
+            buf.writeBytes(NULL_BULK_STRING);
+            buf.writeBytes(CRLF);
+        } else {
+            buf.writeBytes(numberToBytes(msg.children().size()));
+            buf.writeBytes(CRLF);
+            for (RedisMessage child : msg.children()) {
+                writeRedisMessage(allocator, buf, child);
+            }
+        }
+    }
+
+    private static int calculateBufferLength(RedisMessage msg) {
+        if (msg instanceof StringRedisMessage) {
+            // getBytes is not good..
+            return TYPE_LENGTH + ((StringRedisMessage) msg).content().getBytes(CharsetUtil.UTF_8).length + CRLF.length;
+        } else if (msg instanceof IntegerRedisMessage) {
+            return TYPE_LENGTH + getBufferLengthOfNumber(((IntegerRedisMessage) msg).value()) + CRLF.length;
+        } else if (msg instanceof BulkStringRedisMessage) {
+            final BulkStringRedisMessage message = (BulkStringRedisMessage) msg;
+            if (message.isNull()) {
+                return TYPE_LENGTH + NULL_BULK_STRING.length + CRLF.length;
+            } else {
+                final int contentLength = message.content().readableBytes();
+                return TYPE_LENGTH + numberToBytes(contentLength).length + CRLF.length
+                       + contentLength + CRLF.length;
+            }
+        } else if (msg instanceof ArrayHeaderRedisMessage) {
+            return TYPE_LENGTH + getBufferLengthOfNumber(((ArrayHeaderRedisMessage) msg).length()) + CRLF.length;
+        } else if (msg instanceof ArrayRedisMessage) {
+            final ArrayRedisMessage message = (ArrayRedisMessage) msg;
+            if (message.children() == null) {
+                return TYPE_LENGTH + NULL_BULK_STRING.length + CRLF.length;
+            } else {
+                int length = TYPE_LENGTH + getBufferLengthOfNumber(message.children().size()) + CRLF.length;
+                for (RedisMessage child : message.children()) {
+                    length += calculateBufferLength(child);
+                }
+                return length;
+            }
+        } else {
+            throw new CodecException("bad message: " + msg);
+        }
+    }
+
+    private static byte[] numberToBytes(long value) {
+        return Long.toString(value).getBytes(CharsetUtil.US_ASCII);
```

Comment:
This will currently do the following:

char[] -> String -> allocate encoder -> allocate byte[max for encoding] maybe copy into a trimmed byte[] after size is known -> copy that byte[] into ByteBuf.

@normanmaurer - This feels like a good candidate for a followup PR for something like `ByteBufUtil.writeAscii(ByteBuf buf, long value)` ... wdyt?